### PR TITLE
Added field "is_deterministic" to system.functions

### PR DESCRIPTION
**Pull request by rschu1ze**
This is a re-submit of #54766 which was reverted with #55022.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added field "is_deterministic" to system table "system.functions" which indicates whether the result of a function is stable between two invocations (given exactly the same inputs) or not.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit dcdc8bf0e94a971dd7814cb2f0a52beaab33bc00 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55035/dcdc8bf0e94a971dd7814cb2f0a52beaab33bc00/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

**Review by evillique — APPROVED** [review]

**Comment by jrdi** [comment]
@rschu1ze I think that after this PR was merged to master, we started seeing the following errors:

* `DICTIONARIES_WAS_NOT_LOADED`
* `FUNCTION_NOT_ALLOWED`
* `NOT_IMPLEMENTED`
* `SUPPORT_IS_DISABLED`

Checking a summary of those errors, we are not using any of those functions (dictionaries, introspection, transactions, or NLP functions). Everything is working fine, but those errors pollute the logs and make us think there are system issues. Do you think we should rollback this feature?

```
Row 1:
──────
server_shard:       1
version:            23.9.1.1753
name:               DICTIONARIES_WAS_NOT_LOADED
code:               156
quantity:           10
last_error_message: Embedded dictionaries were not loaded. You need to check configuration file.
traceback:          DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool)
DB::Exception::Exception<char const (&) [77]>(int, char const (&) [77])
std::__1::shared_ptr<DB::IFunctionOverloadResolver> DB::FunctionFactory::adaptFunctionToOverloadResolver<DB::FunctionRegionToPopulation>(std::__1::shared_ptr<DB::Context const>)
DB::FunctionFactory::tryGetImpl(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>) const
DB::FunctionFactory::tryGet(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>) const
DB::StorageSystemFunctions::fillData(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn>>>&, std::__1::shared_ptr<DB::Context const>, DB::SelectQueryInfo const&) const
DB::IStorageSystemOneBlock<DB::StorageSystemFunctions>::read(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long)
DB::IStorage::read(DB::QueryPlan&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long)
DB::InterpreterSelectQuery::executeFetchColumns(DB::QueryProcessingStage::Enum, DB::QueryPlan&)
DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::__1::optional<DB::Pipe>)
DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&)
DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&)
DB::InterpreterSelectWithUnionQuery::execute()
DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)
DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, std::__1::shared_ptr<DB::Context>, std::__1::function<void (DB::QueryResultDetails const&)>, std::__1::optional<DB::FormatSettings> const&, std::__1::function<void (DB::IOutputFormat&)>)
DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::__1::optional<DB::CurrentThread::QueryScope>&)
DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&)
DB::HTTPServerConnection::run()
Poco::Net::TCPServerConnection::start()
Poco::Net::TCPServerDispatcher::run()
Poco::PooledThread::run()
Poco::ThreadImpl::runnableEntry(void*)
Row 2:
──────
server_shard:       1
version:            23.9.1.1753
name:               FUNCTION_NOT_ALLOWED
code:               446
quantity:           4
last_error_message: default: Introspection functions are disabled, because setting 'allow_introspection_functions' is set to 0
traceback:          DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool)
DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>(int, FormatStringHelperImpl<std::__1::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>::type>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&)
bool DB::ContextAccess::checkAccessImplHelper<true, false>(DB::AccessFlags) const
std::__1::shared_ptr<DB::IFunctionOverloadResolver> DB::FunctionFactory::adaptFunctionToOverloadResolver<DB::(anonymous namespace)::FunctionAddressToLine>(std::__1::shared_ptr<DB::Context const>)
DB::FunctionFactory::tryGetImpl(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>) const
DB::FunctionFactory::tryGet(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>) const
DB::StorageSystemFunctions::fillData(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn>>>&, std::__1::shared_ptr<DB::Context const>, DB::SelectQueryInfo const&) const
DB::IStorageSystemOneBlock<DB::StorageSystemFunctions>::read(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long)
DB::IStorage::read(DB::QueryPlan&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long)
DB::InterpreterSelectQuery::executeFetchColumns(DB::QueryProcessingStage::Enum, DB::QueryPlan&)
DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::__1::optional<DB::Pipe>)
DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&)
DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&)
DB::InterpreterSelectWithUnionQuery::execute()
DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)
DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, std::__1::shared_ptr<DB::Context>, std::__1::function<void (DB::QueryResultDetails const&)>, std::__1::optional<DB::FormatSettings> const&, std::__1::function<void (DB::IOutputFormat&)>)
DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::__1::optional<DB::CurrentThread::QueryScope>&)
DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&)
DB::HTTPServerConnection::run()
Poco::Net::TCPServerConnection::start()
Poco::Net::TCPServerDispatcher::run()
Poco::PooledThread::run()
Poco::ThreadImpl::runnableEntry(void*)
Row 3:
──────
server_shard:       1
version:            23.9.1.1753
name:               NOT_IMPLEMENTED
code:               48
quantity:           2
last_error_message: Transactions are not supported
traceback:          DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool)
DB::Exception::Exception<char const (&) [31]>(int, char const (&) [31])
DB::Context::checkTransactionsAreAllowed(bool) const
std::__1::shared_ptr<DB::IFunctionOverloadResolver> DB::FunctionFactory::adaptFunctionToOverloadResolver<DB::(anonymous namespace)::FunctionTransactionOldestSnapshot>(std::__1::shared_ptr<DB::Context const>)
DB::FunctionFactory::tryGetImpl(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>) const
DB::FunctionFactory::tryGet(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>) const
DB::StorageSystemFunctions::fillData(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn>>>&, std::__1::shared_ptr<DB::Context const>, DB::SelectQueryInfo const&) const
DB::IStorageSystemOneBlock<DB::StorageSystemFunctions>::read(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long)
DB::IStorage::read(DB::QueryPlan&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long)
DB::InterpreterSelectQuery::executeFetchColumns(DB::QueryProcessingStage::Enum, DB::QueryPlan&)
DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::__1::optional<DB::Pipe>)
DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&)
DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&)
DB::InterpreterSelectWithUnionQuery::execute()
DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)
DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, std::__1::shared_ptr<DB::Context>, std::__1::function<void (DB::QueryResultDetails const&)>, std::__1::optional<DB::FormatSettings> const&, std::__1::function<void (DB::IOutputFormat&)>)
DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::__1::optional<DB::CurrentThread::QueryScope>&)
DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&)
DB::HTTPServerConnection::run()
Poco::Net::TCPServerConnection::start()
Poco::Net::TCPServerDispatcher::run()
Poco::PooledThread::run()
Poco::ThreadImpl::runnableEntry(void*)
Row 4:
──────
server_shard:       1
version:            23.9.1.1753
name:               SUPPORT_IS_DISABLED
code:               344
quantity:           9
last_error_message: Natural language processing function 'detectLanguageUnknown' is experimental. Set `allow_experimental_nlp_functions` setting to enable it
traceback:          DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool)
DB::Exception::Exception<char const* const&>(int, FormatStringHelperImpl<std::__1::type_identity<char const* const&>::type>, char const* const&)
std::__1::shared_ptr<DB::IFunctionOverloadResolver> DB::FunctionFactory::adaptFunctionToOverloadResolver<DB::FunctionTextClassificationString<DB::CharsetClassificationImpl<true>, DB::NameDetectLanguageUnknown>>(std::__1::shared_ptr<DB::Context const>)
DB::FunctionFactory::tryGetImpl(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>) const
DB::FunctionFactory::tryGet(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>) const
DB::StorageSystemFunctions::fillData(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn>>>&, std::__1::shared_ptr<DB::Context const>, DB::SelectQueryInfo const&) const
DB::IStorageSystemOneBlock<DB::StorageSystemFunctions>::read(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long)
DB::IStorage::read(DB::QueryPlan&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long)
DB::InterpreterSelectQuery::executeFetchColumns(DB::QueryProcessingStage::Enum, DB::QueryPlan&)
DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::__1::optional<DB::Pipe>)
DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&)
DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&)
DB::InterpreterSelectWithUnionQuery::execute()
DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)
DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, std::__1::shared_ptr<DB::Context>, std::__1::function<void (DB::QueryResultDetails const&)>, std::__1::optional<DB::FormatSettings> const&, std::__1::function<void (DB::IOutputFormat&)>)
DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::__1::optional<DB::CurrentThread::QueryScope>&)
DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&)
DB::HTTPServerConnection::run()
Poco::Net::TCPServerConnection::start()
Poco::Net::TCPServerDispatcher::run()
Poco::PooledThread::run()
Poco::ThreadImpl::runnableEntry(void*)
Please review these new errors and fix them
```

**Comment by rschu1ze** [comment]
@jrdi Thanks, I can reproduce this issue. The problem is that whenever an exception is created, internal error statistics are updated which then show up in `system.errors`, even if the exception is internally handled, i.e. suppressed. I think the least invasive fix will be to no update these statistics (I'll try that). As an alternative, we could implement a method to query functions if they are deterministic or not *without* constructing the function - but that will be a bigger rewrite.

File: src/Storages/System/StorageSystemFunctions.cpp [hunk 1]
```diff
@@ -102,6 +118,7 @@ NamesAndTypesList StorageSystemFunctions::getNamesAndTypes()
     return {
         {"name", std::make_shared<DataTypeString>()},
         {"is_aggregate", std::make_shared<DataTypeUInt8>()},
+        {"is_deterministic", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeUInt8>())},
```
Comment (antonio2368) [inline src/Storages/System/StorageSystemFunctions.cpp:121]: why is it nullable?

Comment (rschu1ze) [reply src/Storages/System/StorageSystemFunctions.cpp:121]: See l. 50 and l. 141ff. There are situations in which we cannot determine if a function is deterministic or not. In this case, NULL is displayed.

File: tests/queries/0_stateless/02815_no_throw_in_simple_queries.sh [hunk 2]
```diff
@@ -53,4 +53,9 @@ expect eof
 EOF
 }
 
-run "$CLICKHOUSE_LOCAL"
+run "$CLICKHOUSE_LOCAL --disable_suggestion"
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02815_no_throw_in_simple_queries.sh:56]: This was a mistake.

**Comment by rschu1ze** [comment]
Linking the PR which reverts this PR: https://github.com/ClickHouse/ClickHouse/pull/66630

**Pull request merged**